ccl/changefeedccl: implement `WITH diff` option

nvb opened the pull request:
Fixes #28666.

This commit adds a `WITH diff` option to CREATE CHANGEFEED. When the option is provided, changefeeds publications will include a `before` field, which includes the value of the row before the CDC update was applied.

We are able to implement this efficiently by pushing the option into Rangefeed and performing the scan of the previous value immediately before applying the update in the Raft processing goroutine. #41788 allows us to avoid performing this lookup when `WITH diff` isn't specified for Rangefeeds, so the small (unmeasured) perf hit is strictly opt-in.

nvb committed:
ccl/changefeedccl: re-order avro CDC encoding for faster parsing

This commit re-orders the avro encoding of CDC messages to place fields
in the order: [`before`, `after`, `updated`, `resolved`].

Rumor has it that Materialize has a fast-path if these fields are
ordered like this.

Release note (sql change): The fields in the Avro format for CHANGEFEED
records have been re-ordered to allow for optimized parsing. This is a
backwards compatable change.

nvb committed:
kv: set txn MinTimestamp in SetFixedTimestamp

This was causing tests to stall because a transaction's MinTimestamp
could end up being above its commit timestamp.

Release note: None

nvb committed:
ccl/changefeedccl: implement `WITH diff` option

Fixes #28666.
First commit from #41788.

This commit adds a `WITH diff` option to CREATE CHANGEFEED. When the
option is provided, changefeeds publications will include a `before`
field, which includes the value of the row before the CDC update was
applied.

We are able to implement this efficiently by pushing the option into
Rangefeed and performing the scan of the previous value immediately
before applying the update in the Raft processing goroutine. #41788
allows us to avoid performing this lookup when `WITH diff` isn't
specified for Rangefeeds, so the small (unmeasured) perf hit is
strictly opt-in.

Release note (sql change): CHANGEFEED now supports a `WITH diff` option,
which instructs it to include a `before` field in each publication.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41793)
<!-- Reviewable:end -->

danhhz reviewed (commented):
Okay first pass here

- I'm sciencedog on the MinTimestamp change but I trust you know what you're doing there
- Lets leave the avro reordering out of this PR so it can be considered separately

Testing plan
- Very basic with diff and without diff test
- Add it to nemesis test somehow
- Whatever registration/deregistration edge cases you are most worried about

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-:-LroqBjE-S18qgk-ohUC:b-sitkuy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @nvanbenschoten)

___
*[pkg/ccl/changefeedccl/avro.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpICmJ0UHdEhgN2MgX:-LrpICmJ0UHdEhgN2MgY:brnb1sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/avro.go#L521)):*
> ```Go
> 	}
> 	if opts.afterField {
> 		schema.after = after
> ```

do we need a similar line for before?
___
*[pkg/ccl/changefeedccl/avro.go, line 578 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpIKBm53vrBAp5m3ra:-LrpIKBm53vrBAp5m3rb:bf0ibvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/avro.go#L578)):*
> ```Go
> 				return nil, err
> 			}
> 			native[`before`] = goavro.Union(avroUnionKey(&r.after.avroRecord), beforeNative)
> ```

seems weird that this is still named `r.after`. if they're going to be shared, we should probably rename it
___
*[pkg/ccl/changefeedccl/changefeed.go, line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpJvie6QKp3ELt7UZ5:-LrpJvie6QKp3ELt7UZ6:bqlzbas) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/changefeed.go#L106)):*
> ```Go
> 			}
> 			if r.row.datums == nil {
> 				log.Fatalf(ctx, "TODO DURING REVIEW can this happen?")
> ```

good question, maybe not. AssertionErrorf?
___
*[pkg/ccl/changefeedccl/changefeed.go, line 112 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpJiPKE_dx45ap1O5Q:-LrpJiPKE_dx45ap1O5R:bk5lsjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/changefeed.go#L112)):*
> ```Go
> 			r.row.updated = schemaTimestamp
> 
> 			var nextRow emitEntry
> ```

this block is just asserting that we don't get a second row? let's throw in a comment saying that
___
*[pkg/ccl/changefeedccl/encoder.go, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpL2-c6_UulshoUWxw:-LrpKEqTApN4R7zUZ4lY:b-v4ecyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/encoder.go#L49)):*
> ```Go
> 	// prevDatums is the old value of a changed table row.
> 	prevDatums sqlbase.EncDatumRow
> 	// TODO DURING REVIEW: Do we need this?
> ```

i think so, right? what if a row is deleted then re-inserted. we'd see the tombstone when we looked up the previous value. sounds like a good test case
___
*[pkg/ccl/changefeedccl/encoder.go, line 107 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpLIxM2X_rw1CQVpky:-LrpLIxM2X_rw1CQVpkz:bv9de3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/encoder.go#L107)):*
> ```Go
> 	_, e.beforeField = opts[optDiff]
> 	if e.beforeField && !e.wrapped {
> 		return nil, errors.Errorf(`%s is only usable with %s=%s`,
> ```

note to self: make sure there's a test for this
___
*[pkg/ccl/changefeedccl/encoder.go, line 435 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpLoGw3KR8ReUXNq-3:-LrpLoGw3KR8ReUXNq-4:brdwu53) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/encoder.go#L435)):*
> ```Go
> 	}
> 	binary.BigEndian.PutUint32(header[1:5], uint32(registered.registryID))
> 	return registered.schema.BinaryFromRow(header, meta, nil /* row */, nil)
> ```

nit: comment for second nil
___
*[pkg/ccl/changefeedccl/poller.go, line 524 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpM5rv75shbT7SxUee:-LrpM5rv75shbT7SxUef:bihag2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/poller.go#L524)):*
> ```Go
> 		sort.Sort(byValueTimestamp(kvs))
> 		for _, kv := range kvs {
> 			if err := p.buf.AddKV(ctx, kv, nil /* prevVal */, schemaTimestamp); err != nil {
> ```

oh hmmmm this is a good question. what should be the value for the backfill we run after schema changes? leaving it unset is a little strange
___
*[pkg/roachpb/api.proto, line 1847 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpP93h4zkswrhQux2p:-LrpOXJW2vcOvCPSus5L:b-b2lmhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/roachpb/api.proto#L1847)):*
> ```protobuf
>   bytes key       = 1 [(gogoproto.casttype) = "Key"];
>   Value value     = 2 [(gogoproto.nullable) = false];
>   // TODO DURING REVIEW: how should this be represented? Should this be:
> ```

okay, so the two axis here are a) is the consumer responsible for knowing whether the with_diff option was set on the request and b) does the consumer need to know the difference between the previous key is a deletion vs it never existed. 1 is no and yes, 2 is no and no, 3 is yes and no

a) i could go either way
b) where do we currently expose the difference between a key that was deleted and one that never existed? at the moment, i think it's basically only things that involve spans of time (incremental backup, rangefeed catchup scan). these events are all single point in time, so i think i lean toward not needed to differentiate between these

so....given that non-nullable fields are generally preferred, i guess 3, which is what you already have. i don't feel like i have a strong feel for this though


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Testing plan
> 
> Very basic with diff and without diff test
> Add it to nemesis test somehow
> Whatever registration/deregistration edge cases you are most worried about

Done.
Done.
Done.

The only remaining things missing here are addressing some review feedback and getting this working with schema changes. I'm going to sit on those items until #41842 lands so I'm not stepping on @aayushshah15's toes.

@benesch this should be ready for use with TPC-CH.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-:-Lryl-bD5MCUBx9ULwtO:bib3bn1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benesch, @danhhz, and @nvanbenschoten)

___
*[pkg/ccl/changefeedccl/avro.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpICmJ0UHdEhgN2MgX:-LrylMEeAv4UHsy5ZYV9:b-yeywb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/avro.go#L521)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

do we need a similar line for before?
</blockquote></details>

It's a little unclear to me what we should do here in the cases where the `before` and `after` fields have a) the same schema b) different schemas. In case a), we could either define two schemas with _different names_ or we could define the schema in one place and reference it in both. In case b) we will need to define two schemas with _different names_. Do you have a preference here? Or maybe @benesch has insight into the right approach?
___
*[pkg/ccl/changefeedccl/avro.go, line 578 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpIKBm53vrBAp5m3ra:-LryltR72PnZESCeTiE6:bbynie7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/avro.go#L578)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

seems weird that this is still named `r.after`. if they're going to be shared, we should probably rename it
</blockquote></details>

Agreed, I'll rename.
___
*[pkg/ccl/changefeedccl/encoder.go, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpL2-c6_UulshoUWxw:-LrylzsxDU7skTOSE5C3:b-xlazpg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/encoder.go#L49)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

i think so, right? what if a row is deleted then re-inserted. we'd see the tombstone when we looked up the previous value. sounds like a good test case
</blockquote></details>

This ties into the question we're discussing below.
___
*[pkg/ccl/changefeedccl/encoder.go, line 107 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpLIxM2X_rw1CQVpky:-Lrym3mo6TfULB2dD0Fc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/encoder.go#L107)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

note to self: make sure there's a test for this
</blockquote></details>

Done.
___
*[pkg/roachpb/api.proto, line 1847 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpP93h4zkswrhQux2p:-Lryml9v398CBvV5VMqS:b-93dsfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/roachpb/api.proto#L1847)):*
> a) is the consumer responsible for knowing whether the with_diff option was set on the request

Yeah, I don't have a strong feeling about this either. I'd lean towards no.

> b) does the consumer need to know the difference between the previous key is a deletion vs it never existed

I'd also say no on this one. I don't think `prevValue` should ever be a deletion tombstone.

So that points me towards 3: non-nullable and only populated if prev val exists and `with_diff=true` specified.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benesch, @danhhz, and @nvanbenschoten)

___
*[pkg/ccl/changefeedccl/avro.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpICmJ0UHdEhgN2MgX:-Lryydhc0R-GfoL2zaNG:bppic5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/avro.go#L521)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's a little unclear to me what we should do here in the cases where the `before` and `after` fields have a) the same schema b) different schemas. In case a), we could either define two schemas with _different names_ or we could define the schema in one place and reference it in both. In case b) we will need to define two schemas with _different names_. Do you have a preference here? Or maybe @benesch has insight into the right approach?
</blockquote></details>

I also am having trouble developing a strong opinion here. @benesch if you have one it would be welcomed


<!-- Sent from Reviewable.io -->

benesch reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @nvanbenschoten)

___
*[pkg/ccl/changefeedccl/avro.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpICmJ0UHdEhgN2MgX:-LsIY9gnAXWaaKSFWVAX:b-w7rnyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/avro.go#L521)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I also am having trouble developing a strong opinion here. @benesch if you have one it would be welcomed
</blockquote></details>

Forgive me if I'm being dense, but... how could `before` and `after` have different schemas?


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benesch, @danhhz, and @nvanbenschoten)

___
*[pkg/ccl/changefeedccl/avro.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpICmJ0UHdEhgN2MgX:-LsIYnw83rU2PM6yHg-i:bgqe6qf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/avro.go#L521)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Forgive me if I'm being dense, but... how could `before` and `after` have different schemas?
</blockquote></details>

If the timestamp of the row change exactly equals the timestamp of a schema change, then the before will have the previous schema. Perhaps surprisingly, this is not a rare edge case. It happens for every schema change because we do the changefeed-level backfill at exactly the time the column shows up/is removed from the SQL perspective.

(Hmm, I think materialize might want to skip the changefeed-level backfill entirely, which is an option I've always assumed we'd have at some point but we've never implemented. Anyway, you seem to have more experience with avro schema registry than me at this point, so I'd still like your take)


<!-- Sent from Reviewable.io -->

benesch reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @nvanbenschoten)

___
*[pkg/ccl/changefeedccl/avro.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpICmJ0UHdEhgN2MgX:-LsIdel95frB-8oxy8kC:b-je4t3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/avro.go#L521)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

If the timestamp of the row change exactly equals the timestamp of a schema change, then the before will have the previous schema. Perhaps surprisingly, this is not a rare edge case. It happens for every schema change because we do the changefeed-level backfill at exactly the time the column shows up/is removed from the SQL perspective.

(Hmm, I think materialize might want to skip the changefeed-level backfill entirely, which is an option I've always assumed we'd have at some point but we've never implemented. Anyway, you seem to have more experience with avro schema registry than me at this point, so I'd still like your take)
</blockquote></details>

🤯

I guess that makes sense! Let me see what Debezium does, if anything. The Debezium docs claim that the before schema is always equal to the after schema, and it would stink not to have that invariant. But I'm also not seeing another sensible way to represent a change event that spans a schema change, so perhaps the Debezium docs are just wrong in the face of schema changes.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benesch, @danhhz, and @nvanbenschoten)

___
*[pkg/ccl/changefeedccl/avro.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpICmJ0UHdEhgN2MgX:-LsIiNtz2k54hDTOWurO:b-e5jepp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/avro.go#L521)):*
>  an option I've always assumed we'd have at some point but we've never implemented

https://github.com/cockroachdb/cockroach/issues/31213 for posterity 


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Rebased on top of #41842. As expected `TestChangefeedNemeses` now fails due to schema change events. If I disable those events, it passes again. This lack of proper support for schema changes is also picked up by the `cdc/schemareg` roachtest.

Getting schema changes working here is the last remaining item before this is ready to go. For that, we'll need to answer how a schema change should influence the schema of the `before` field in Avro records.

lopezator commented:
Is there any advance on this? We are interested on this feature. Thanks!

nvanbenschoten commented:
Hi @lopezator! I'm thrilled to hear that you're interested in the feature. This will, in all likelihood, be making it into the next release of Cockroach. We just released v19.2 last week, so we're probably about 6 months out from that, but the plan is to begin publishing v20.1 alphas shortly. If you're willing to test on alpha releases (or even on master), we can probably get it into your hands fairly soon. We'd love to hear your feedback on it once you're able to test it out.

@danhhz This has been updated to support schema changes correctly, as we discussed. When a schema change triggers a backfill, it produces a record with the same KV pair interpreted with the old and new schema as the "before" and "after" fields, respectively. This is exactly the contract that SQL enforces when performing AOST queries at and immediately before the schema change timestamp, as is now tested by the `beforeAfterValidator`. I think this is ready for a final review.

lopezator commented:
@nvanbenschoten that are good news!

We are cockroach cloud users, but possibly wouldn't mind to update the cluster to an alpha or a master checkout to bring this feature in.

By the way, thanks a lot for your great work.

CC/ @glerchundi

danhhz reviewed (commented):
Okay, I think this is getting really close. Everything here is a nit except two larger things
1) The logic to switch between whether it's the initial backfill (no previous values) or a schema change backfill (yes previous values) seems off. If so, this is a hole in the testing. I think the nemesis test is not catching it because BeforeAfterValidator is not strictly asserting that the first time it sees a key, that before should always be unset. 
2) I'm confused about our treatment of the previous value's timestamp and which schema we use to print it. If this is correct, it's not clear to me why and we should put a mini-overview somewhere. Even if it's correct now, it seems to me that it would be much more straightforward if the previous value's timestamp was threaded all the way from rangefeed, but perhaps I'm wrong here? Curious what your take is on this

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-:-LtzcwI69bydiA1SHTYZ:b-q75oq8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benesch, @danhhz, and @nvanbenschoten)

___
*[pkg/ccl/changefeedccl/avro.go, line 387 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzRiM_D10Tf2RPEa5U:-LtzRiM_D10Tf2RPEa5V:b-v1io99) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/avro.go#L387)):*
> ```Go
> 	name := SQLNameToAvroName(tableDesc.Name)
> 	if nameSuffix != "" {
> 		name = fmt.Sprintf("%s_%s", name, nameSuffix)
> ```

nit: isn't it faster (and clearer) to + the strings here?
___
*[pkg/ccl/changefeedccl/avro.go, line 606 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzS-Pb5K7ssotw_f1z:-LtzS-Pb5K7ssotw_f2-:bkkfmmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/avro.go#L606)):*
> ```Go
> 		}
> 	}
> 	// WIP verify that meta is now empty
> ```

would you mind doing this WIP while you're in here? should just be something like```if k := range meta { return nil, errors.AssertionErrorf(`unhandled meta key: %s`, k) }```
___
*[pkg/ccl/changefeedccl/avro_test.go, line 119 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzSi4NDgMoXYm8RBj7:-LtzSi4NDgMoXYm8RBj8:b-lmgy8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/avro_test.go#L119)):*
> ```Go
> 		tableDesc.Columns = append(tableDesc.Columns, *colDesc)
> 	}
> 	return tableToAvroSchema(tableDesc, "")
> ```

thoughts on a named constant for this? avroSchemaNoSuffix?
___
*[pkg/ccl/changefeedccl/buffer.go, line 30 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzSv5tDq45oZAkc7sb:-LtzSv5tDq45oZAkc7sc:b-qgrv4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/buffer.go#L30)):*
> ```Go
> type bufferEntry struct {
> 	kv       roachpb.KeyValue
> 	prevVal  *roachpb.Value
> ```

comment explaining when this is set/not set pls
___
*[pkg/ccl/changefeedccl/buffer.go, line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzT1huEUDWh-2k4tRV:-LtzT1huEUDWh-2k4tRW:b1orc38) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/buffer.go#L37)):*
> ```Go
> 	// Timestamp of the schema that should be used to read the previous
> 	// version of this KV.
> 	// If unset (zero-valued), the previous value will be interpretted with
> ```

instead of zero value having a special meaning, i'd find it more straightforward if we fill it in whenever prevVal is set

edit: actually i have larger questions about previous value timestamps and schemas
___
*[pkg/ccl/changefeedccl/buffer.go, line 111 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzYyiaDjkimX3mv3_2:-LtzYyiaDjkimX3mv3_3:b-79cok0) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/buffer.go#L111)):*
> ```Go
> 	*types.Bytes, // span.Key
> 	*types.Bytes, // span.EndKey
> 	*types.Int,   // ts.WallTime
> ```

I'm confused about why we don't need to pass the prevVal's timestamp through here
___
*[pkg/ccl/changefeedccl/changefeed.go, line 112 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpJiPKE_dx45ap1O5Q:-LtzU-WX74aND1ejbyDX:b-zctf15) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/changefeed.go#L112)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this block is just asserting that we don't get a second row? let's throw in a comment saying that
</blockquote></details>

ping : - )
___
*[pkg/ccl/changefeedccl/changefeed.go, line 125 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzU7RuDt6riTBouxMo:-LtzU7RuDt6riTBouxMp:b-xh4c83) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/changefeed.go#L125)):*
> ```Go
> 			}
> 			if nextRow.row.datums != nil {
> 				return nil, errors.AssertionFailedf("unexpected non-empty datums")
> ```

is there an easy way to print the datums in this message?
___
*[pkg/ccl/changefeedccl/changefeed.go, line 133 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzUFFgDezDqRrHIS8I:-LtzUFFgDezDqRrHIS8J:bbwjhua) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/changefeed.go#L133)):*
> ```Go
> 			prevRF := rf
> 			if prevSchemaTimestamp != schemaTimestamp {
> 				// If the previous value is being interpretted under a different
> ```

nit: interpreted
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 1935 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzV-7B3g2ZROhkxe5v:-LtzV-7B3g2ZROhkxe5w:b-l6mwxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/changefeed_test.go#L1935)):*
> ```Go
> func TestChangefeedRestartDuringBackfill(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	// t.Skip("TODO(WIP): doesn't work yet")
> ```

remove
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 2027 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzVJ_14_jVfORfK-AJ:-LtzVJ_14_jVfORfK-AK:b-qb6q6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/changefeed_test.go#L2027)):*
> ```Go
> 			// TODO(dan): Track duplicates more precisely in sinklessFeed/tableFeed.
> 			// `foo: [0]->{"after": {"a": 0, "b": "backfill"}}`,
> 			// TODO DURING REVIEW: is this correct? Should the backfill on a RESUME JOB
> ```

looks correct to me. it was paused during a schema change backfill and this test is verifying that it restarts the schema change backfill when unpaused
___
*[pkg/ccl/changefeedccl/encoder.go, line 435 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpLoGw3KR8ReUXNq-3:-LtzWx1lCfEd7pLtodfY:b-zctf15) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/encoder.go#L435)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: comment for second nil
</blockquote></details>

ping : - )
___
*[pkg/ccl/changefeedccl/encoder.go, line 52 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzVdH-0aaQOsEmg0q5:-LtzVdH-0aaQOsEmg0q6:b-8dyte1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/encoder.go#L52)):*
> ```Go
> 	// prevDatums is the old value of a changed table row.
> 	prevDatums sqlbase.EncDatumRow
> 	// TODO DURING REVIEW: Do we need this?
> ```

yes, this would happen if we inserted a previously deleted value. we should add a test for this if there isn't one already
___
*[pkg/ccl/changefeedccl/encoder.go, line 374 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzWFW10N5YOUWeOZkN:-LtzWFW10N5YOUWeOZkO:bt7sq1t) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/encoder.go#L374)):*
> ```Go
> 		}
> 
> 		afterDataSchema, err := tableToAvroSchema(row.tableDesc, ``)
> ```

my suggestion offline was for this suffix to be `after`. however, i like that this is the same name it had previously, so i'm okay with this. i wish there was better precedent here. i wonder if @rolandcrosby knows any
___
*[pkg/ccl/changefeedccl/encoder_test.go, line 131 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzXJLz39tPtlIHfbs8:-LtzXJLz39tPtlIHfbs9:b-uqow9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/encoder_test.go#L131)):*
> ```Go
> 			resolved: `{"resolved":{"string":"1.0000000002"}}`,
> 		},
> 		`format=experimental_avro,envelope=key_only,diff`: {
> ```

this should error with the same thing as `format=json,envelope=row,diff` right?
___
*[pkg/ccl/changefeedccl/poller.go, line 191 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzXzQ73AtahA2--8b9:-LtzXzQ73AtahA2--8bA:b-8jmapx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/poller.go#L191)):*
> ```Go
> 	// Perform a full scan if necessary - either an initial scan or a backfill
> 	// Full scans are still performed using an Export operation.
> 	initialScan := i == 0
> ```

is this logic correct when resuming? i'd think it should be based on the frontier
___
*[pkg/storage/rangefeed/registry.go, line 121 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzQvH25NrswUvNddS5:-LtzQvH25NrswUvNddS6:buaja6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/storage/rangefeed/registry.go#L121)):*
> ```Go
> 	case *roachpb.RangeFeedValue:
> 		if t.Key == nil {
> 			panic(fmt.Sprintf("unexpected empty RangeFeedValue.Key: %v", t))
> ```

I wonder if we do anything special if this panic gets an assertion error instead of a string


<!-- Sent from Reviewable.io -->

nvb committed:
ccl/changefeedccl: address TODO in BinaryFromRow

We now verify that all fields in the avroMetadata are encoded.

nvb reviewed (commented):
TFTR!

> I think the nemesis test is not catching it because BeforeAfterValidator is not strictly asserting that the first time it sees a key, that before should always be unset.

See my comment below. The one that starts with "I'm not really sure".

> it's not clear to me why and we should put a mini-overview somewhere

Completely agree. Does some overview exist that I can add to? I hope I missed it because it was pretty tough to track down the flow of things through the tableHistory's validate function through the poller's `scanBoundaries`, which eventually triggered the backfill (do I have that right?).

> it seems to me that it would be much more straightforward if the previous value's timestamp was threaded all the way from rangefeed, but perhaps I'm wrong here

Well, the previous value's timestamp never comes from Rangefeed, and it never needs to. The previous value is logically set right up to the new value's timestamp. I guess we could always treat the previous value's timestamp as `curVal.Timestamp.Prev()` and always look up the schema at that timestamp. Then we wouldn't need any special logic when doing a backfill. That would lead to wasted work in the overwhelmingly common case where we know the schemas to be identical though. What do you think?

EDIT: actually, it seems like we're in the exact same boat with this `schemaTimestamp`. It's only set on backfills. In fact, it doesn't even need to be passed through `memBuffer` (want me to add a commit?). Can we unify this with `prevSchemaTimestamp` such that if schemaTimestamp is provided we treat `prevVal.Timestamp` as `schemaTimestamp.Prev` and if not we treat `prevVal.Timestamp` as `curVal.Timestamp.Prev`?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-:-Lu-wf2a7aSScNVMsl7K:b-f6zd0h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @rolandcrosby)

___
*[pkg/ccl/changefeedccl/avro.go, line 387 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzRiM_D10Tf2RPEa5U:-Lu-dYDhED_H0BXWD_RJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/avro.go#L387)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: isn't it faster (and clearer) to + the strings here?
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/avro.go, line 606 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzS-Pb5K7ssotw_f1z:-Lu-eJHsD-TRyyTlmJ2s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/avro.go#L606)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

would you mind doing this WIP while you're in here? should just be something like```if k := range meta { return nil, errors.AssertionErrorf(`unhandled meta key: %s`, k) }```
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/avro_test.go, line 119 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzSi4NDgMoXYm8RBj7:-Lu-eaTG0ffoFopiYnHE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/avro_test.go#L119)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

thoughts on a named constant for this? avroSchemaNoSuffix?
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/buffer.go, line 30 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzSv5tDq45oZAkc7sb:-Lu-tjQv6GWm1YwgVXy1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/buffer.go#L30)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

comment explaining when this is set/not set pls
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/buffer.go, line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzT1huEUDWh-2k4tRV:-Lu-vQYtCjGyDlpbNp6C:b-xt8yty) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/buffer.go#L37)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

instead of zero value having a special meaning, i'd find it more straightforward if we fill it in whenever prevVal is set

edit: actually i have larger questions about previous value timestamps and schemas
</blockquote></details>

See below. In almost all cases, the two timestamps are the same. It's only in the schema change case where the timestamps diverge.
___
*[pkg/ccl/changefeedccl/buffer.go, line 111 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzYyiaDjkimX3mv3_2:-Lu-uuHV0nrjZrJ1vVVb:b-nmd53e) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/buffer.go#L111)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I'm confused about why we don't need to pass the prevVal's timestamp through here
</blockquote></details>

Because the only time that we need to provide it is when we're performing a schema change backfill, in which case we skip the `memBuffer`. On the normal path, the prevVal timestamp isn't provided.
___
*[pkg/ccl/changefeedccl/changefeed.go, line 112 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpJiPKE_dx45ap1O5Q:-Lu-fezTA4-NwNrz6LrQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/changefeed.go#L112)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

ping : - )
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed.go, line 125 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzU7RuDt6riTBouxMo:-Lu-fhMQAY7r2Slt-pnY:b-djyhr4) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/changefeed.go#L125)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

is there an easy way to print the datums in this message?
</blockquote></details>

Not really. I had some debug information around here for a different reason and we had to use the schema to decode the datums. I think it's more effort than it's worth for this assertion. It really should never fire because we gave the row fetcher a single KV.
___
*[pkg/ccl/changefeedccl/changefeed.go, line 133 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzUFFgDezDqRrHIS8I:-Lu-g-KwAwVF83OqYUzX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/changefeed.go#L133)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: interpreted
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 1935 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzV-7B3g2ZROhkxe5v:-Lu-g7VO7IZI_9W9ohBa:b-gih2qj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/changefeed_test.go#L1935)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

remove
</blockquote></details>

Woops. Done.
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 2027 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzVJ_14_jVfORfK-AJ:-Lu-gEZ35d5jmi_dY4kX:b513d5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/changefeed_test.go#L2027)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

looks correct to me. it was paused during a schema change backfill and this test is verifying that it restarts the schema change backfill when unpaused
</blockquote></details>

:thumbsup: thanks for the explanation. Removed.
___
*[pkg/ccl/changefeedccl/encoder.go, line 435 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpLoGw3KR8ReUXNq-3:-Lu-oTIaC1PRqUqivf2y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/encoder.go#L435)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

ping : - )
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/encoder.go, line 52 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzVdH-0aaQOsEmg0q5:-Lu-o1HuAKxjrWqjqqbH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/encoder.go#L52)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

yes, this would happen if we inserted a previously deleted value. we should add a test for this if there isn't one already
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/encoder.go, line 374 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzWFW10N5YOUWeOZkN:-Lu-o3NH2VeSmbJIbtrw:bso80yw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/encoder.go#L374)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

my suggestion offline was for this suffix to be `after`. however, i like that this is the same name it had previously, so i'm okay with this. i wish there was better precedent here. i wonder if @rolandcrosby knows any
</blockquote></details>

Yeah, I had `after` and then removed it because it broke tests and generally pointed at backwards incompatibility. I ended up liking this approach more.
___
*[pkg/ccl/changefeedccl/encoder_test.go, line 131 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzXJLz39tPtlIHfbs8:-Lu-qj5V4Sx5lmEYqlyu:bylxkh1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/encoder_test.go#L131)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this should error with the same thing as `format=json,envelope=row,diff` right?
</blockquote></details>

Yes, good catch. Done.

Do we need the same error for `format=experimental_avro,envelope=key_only,updated`?
___
*[pkg/ccl/changefeedccl/poller.go, line 524 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpM5rv75shbT7SxUee:-Lu-sLzo3z39laWycal-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/ccl/changefeedccl/poller.go#L524)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

oh hmmmm this is a good question. what should be the value for the backfill we run after schema changes? leaving it unset is a little strange
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/poller.go, line 191 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzXzQ73AtahA2--8b9:-Lu-w-rBApLZPfZJLsnw:bw7zgcw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/poller.go#L191)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

is this logic correct when resuming? i'd think it should be based on the frontier
</blockquote></details>

I'm not really sure. I'm going to need some help understanding the different cases at play here. Is it correct that we perform backfills in these three cases:
1. upon job creation
2. upon job resumption
3. on schema changes

If so then which of these do we expect to include `before` fields? Definitely the third, but also the second? What does `before` mean in that case? I think these semantics need a bit more refinement.
___
*[pkg/roachpb/api.proto, line 1847 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LrpP93h4zkswrhQux2p:-Lu-sbpSClRVKlAlTw7A:b-c1cjhs) ([raw file](https://github.com/cockroachdb/cockroach/blob/5acba54605c5a774e6bc602a38c018df37d03a63/pkg/roachpb/api.proto#L1847)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> a) is the consumer responsible for knowing whether the with_diff option was set on the request

Yeah, I don't have a strong feeling about this either. I'd lean towards no.

> b) does the consumer need to know the difference between the previous key is a deletion vs it never existed

I'd also say no on this one. I don't think `prevValue` should ever be a deletion tombstone.

So that points me towards 3: non-nullable and only populated if prev val exists and `with_diff=true` specified.
</blockquote></details>

Switched to approach 3. It turned out clean.
___
*[pkg/storage/rangefeed/registry.go, line 121 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzQvH25NrswUvNddS5:-Lu-sXnY-iTaeBnKcRks:bg29mml) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/storage/rangefeed/registry.go#L121)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I wonder if we do anything special if this panic gets an assertion error instead of a string
</blockquote></details>

I don't know of anything. Are you aware of special handling?


<!-- Sent from Reviewable.io -->

ajwerner commented:
@aayushshah15 do you want to take a look at this? Especially as it pertains to https://github.com/cockroachdb/cockroach/pull/42053.

aayushshah15 commented:
I've been passively following the PR, don't think my change should clash with this. Though I think I could wait until this is in, rebase and `stress` my change for a little bit.

danhhz reviewed (commented):
Thinking out loud here, there are a few timestamps involved. Listing them out explicitly (some of these end up being the same):

- tUpdated: the timestamp in the updates field of an emitted message
- tAfterKV: the timestamp of the kv being emitted as a row
- tAfterSchema: the timestamp used to select a tabledesc for interpreting the after
- tBeforeKV: the timestamp of the previous value
- tBeforeSchema: the timestamp used to select a tabledesc for interpreting the before
- tBackfill: the timestamp a backfill is being run at

For the initial backfill when a changefeed starts up:
- tAfterKV and tAfterSchema are set to tBackfill
- tBeforeKV, tBeforeSchema are non-applicable, there are no before values in the initial backfill

During normal operation:
- tAfterSchema and tUpdated is set to tAfterKV
- tBeforeSchema is set to tAfterKV (see below for me re-convincing myself of this)

During a schema change backfill:
- tAfterKV and tAfterSchema are set to tBackfill
- tUpdated is set to tBackfill
- tBeforeSchema is set to tBackfill.Prev()

You are correct that nothing depends on tBeforeKV. My remaining question is where all the "is set to" logic lives. Right now it seems haphazard, especially because of the name "schemaTimestamp". I propose that `bufferEntry` have two timestamps: one in the kv and schemaTimestamp is renamed to backfillTimestamp. Then the "is set to" logic is entirely moved to the block in`kvsToRows` where it mostly already is: https://github.com/cockroachdb/cockroach/blob/5765fd2e2c1c84b6abadb336224311842388e944/pkg/ccl/changefeedccl/changefeed.go#L196-L203

It's still possible to remove the second timestamp (schemaTimestamp/backfillTimestamp) from memBuffer.

Thoughts?



---

# Justification for "tBeforeSchema is set to tAfterKV" during normal operation

It's possible you and I already hashed the following out and I forgot, but I had to re-convince myself this morning of what timestamp we should be using when grabbing a schema for the before. Given the following situation:

```
CREATE TABLE foo (a INT PRIMARY KEY)
CREATE CHANGEFEED FOR TABLE foo ... WITH diff
INSERT INTO foo VALUES (1)
ALTER TABLE foo ADD COLUMN b STRING DEFAULT 'default' (gets a backfill)
ALTER TABLE foo ADD COLUMN c STRING (does not get a backfill)
UPDATE foo SET c='update' WHERE a = 1
```

Initially, I see two ways this could work.

1) For each before, use the schema that was active at the time the _before_ was written. In this world, each before for a row matches the non-backfill after that precedes it. So

    ```
    before=null after=(1)
    before=(1) after=(1, 'default') (the backfill)
    before=(1) after=(1, 'default', 'update') (matches the previous non-backfill row)
    ```

    This is mostly interesting because everything lines up when we add the option to skip backfills.

    ```
    before=null after=(1)
    before=(1) after=(1, 'default', 'update')
    ````

2) For each before, use the schema that's active at the time the _after_ was written. This gets us the same answer as if we'd done an AS OF SYSTEM TIME at the time immediately before the update timestamp.

    ```
    before=null after=(1)
    before=(1) after=(1, 'default') (the backfill)
    before=(1, 'default', null) after=(1, 'default', 'update')
    ```
    
When I made that last round of comments, I was thinking of (1) but what you have is (2) and I think that's right, so disregard what I was saying about this in the last round of comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-:-Lu3GiOsBzU6fFGmrJ4P:b-58bxln)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz, @nvanbenschoten, and @rolandcrosby)

___
*[pkg/ccl/changefeedccl/changefeed.go, line 125 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzU7RuDt6riTBouxMo:-Lu3EIe67xWKZUcfgvWq:b-q8ilnv) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/changefeed.go#L125)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Not really. I had some debug information around here for a different reason and we had to use the schema to decode the datums. I think it's more effort than it's worth for this assertion. It really should never fire because we gave the row fetcher a single KV.
</blockquote></details>

gotcha
___
*[pkg/ccl/changefeedccl/encoder_test.go, line 131 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzXJLz39tPtlIHfbs8:-Lu3Egm02mpoHltcsJV0:b-496d8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/encoder_test.go#L131)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yes, good catch. Done.

Do we need the same error for `format=experimental_avro,envelope=key_only,updated`?
</blockquote></details>

Oh, good catch, yes. Though in contrast to diff, which only works with env=wrapped, updated will work with either env=wrapped or env=row. This is technically a backward incompatible change, so it should be it's own commit with release note. I've you don't feel like picking that off, i'm fine with us filing it as an issue for followup.
___
*[pkg/ccl/changefeedccl/poller.go, line 191 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzXzQ73AtahA2--8b9:-Lu3F5jp3dVI4mYph8PW:biw7bq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/poller.go#L191)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm not really sure. I'm going to need some help understanding the different cases at play here. Is it correct that we perform backfills in these three cases:
1. upon job creation
2. upon job resumption
3. on schema changes

If so then which of these do we expect to include `before` fields? Definitely the third, but also the second? What does `before` mean in that case? I think these semantics need a bit more refinement.
</blockquote></details>

We don't backfill on (2) unless it was paused during a type (1) or (3) backfill, in which case it has to resume the backfill it had started but not finished. Maybe that's where your confusion is coming from?

This combined with your question above about where we'd put a comment with details on the timestamp flow makes me think we're long overdue for a changefeed mega-comment. I'll try to circle back and do that at some point.
___
*[pkg/storage/rangefeed/registry.go, line 121 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzQvH25NrswUvNddS5:-Lu3G2Sv4QfBpQ3Q-5zI:b-1tfety) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/storage/rangefeed/registry.go#L121)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I don't know of anything. Are you aware of special handling?
</blockquote></details>

I think we at least get the type information in sentry. The string makes it look like an assertion, so perhaps the string is fine. I guess I lean slightly toward panic'ing an AssertionErrorf here, but I don't feel strongly


<!-- Sent from Reviewable.io -->

nvb committed:
changefeedccl: reject `updated` option when using Avro's `key_only` envelope

Before this change, we would accept the `WITH updated` option when using Avro's
`WITH envelope=key_only` option but would never actually output any updated
fields. This commit properly rejects the config, which doesn't make any sense.

Release note (backwards-incompatible change): The combination of
CHANGEFEED's format=experimental_avro, envelope=key_only, and updated
is now rejected. This is because the use of key_only prevents any
rows with updated fields from being emitted, so the updated option
is meaningless.

nvb committed:
changefeedccl: rework backfillTimestamp handling

This change cleans up the handling of bufferEntry.backfillTimestamp
and removes the schemaTimestamp from memBuffer entirely.

nvb reviewed (commented):
> When I made that last round of comments, I was thinking of (1) but what you have is (2) and I think that's right, so disregard what I was saying about this in the last round of comments.

Yes, we have (2) in this PR, and AS OF SYSTEM TIME is exactly what the `beforeAfterValidator` checks.

>It's still possible to remove the second timestamp (schemaTimestamp/backfillTimestamp) from memBuffer.
> Thoughts?

I like that all. I added a new commit that reworks the handling of these timestamps and I think it came out pretty clean.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-:-Lu58DlW6WRU2yLKNe-a:b-3fg5d1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz)

___
*[pkg/ccl/changefeedccl/encoder_test.go, line 131 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzXJLz39tPtlIHfbs8:-Lu56Fcx61zenxM-ZpqO:b-uq6zt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/encoder_test.go#L131)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Oh, good catch, yes. Though in contrast to diff, which only works with env=wrapped, updated will work with either env=wrapped or env=row. This is technically a backward incompatible change, so it should be it's own commit with release note. I've you don't feel like picking that off, i'm fine with us filing it as an issue for followup.
</blockquote></details>

Done in a new commit.
___
*[pkg/ccl/changefeedccl/poller.go, line 191 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzXzQ73AtahA2--8b9:-Lu5BUIe6kfqnIcSpJhx:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/poller.go#L191)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

We don't backfill on (2) unless it was paused during a type (1) or (3) backfill, in which case it has to resume the backfill it had started but not finished. Maybe that's where your confusion is coming from?

This combined with your question above about where we'd put a comment with details on the timestamp flow makes me think we're long overdue for a changefeed mega-comment. I'll try to circle back and do that at some point.
</blockquote></details>

:thumbsup: 
___
*[pkg/storage/rangefeed/registry.go, line 121 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzQvH25NrswUvNddS5:-Lu575Mp5sTQ7Bw1OH1n:b-fpp8uf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/storage/rangefeed/registry.go#L121)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I think we at least get the type information in sentry. The string makes it look like an assertion, so perhaps the string is fine. I guess I lean slightly toward panic'ing an AssertionErrorf here, but I don't feel strongly
</blockquote></details>

I'm going to leave this because there's not a strong reason to change it and this PR is large enough.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> :clap:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-:-Lu9c42n27AMv-AUDDMw:b-ggulid)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz)

___
*[pkg/ccl/changefeedccl/encoder.go, line 52 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzVdH-0aaQOsEmg0q5:-Lu9bXt55qY7EKkS6Y4s:b-5qu26g) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/encoder.go#L52)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Sorry if I missed it, but is there a test for this? I think adding an INSERT after the DELETE in TestChangefeedDiff would be sufficient


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
I stressed `TestChangefeedNemeses` for 30 minutes after rebasing on top of #42053. Everything looks good. TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-:-LuECNa37rfF9IzuLsGo:b-fw0tnv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz)

___
*[pkg/ccl/changefeedccl/encoder.go, line 52 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/41793#-LtzVdH-0aaQOsEmg0q5:-LuE1Yz195e-3_u7Z8GS:b-cfhoe1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5de4b9a83fe62d68d3796407a83a24c44a52fa/pkg/ccl/changefeedccl/encoder.go#L52)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Sorry if I missed it, but is there a test for this? I think adding an INSERT after the DELETE in TestChangefeedDiff would be sufficient
</blockquote></details>

Yes, without this a number of tests fail. Adding an INSERT after the DELETE in TestChangefeedDiff is a good idea though. Done.


<!-- Sent from Reviewable.io -->
